src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -150,6 +152,82 @@ CorProfiler::Initialize(IUnknown* cor_profiler_info_unknown) {
   return S_OK;
 }
 
+HRESULT STDMETHODCALLTYPE CorProfiler::AssemblyLoadFinished(AssemblyID assembly_id,
+    HRESULT hr_status) {
+  if (FAILED(hr_status)) {
+    // if assembly failed to load, skip it entirely,
+    // otherwise we can crash the process if module is not valid
+    CorProfilerBase::AssemblyLoadFinished(assembly_id, hr_status);
+    return S_OK;
+  }
+
+  if (!is_attached_) {
+    return S_OK;
+  }
+
+  if (debug_logging_enabled) {
+    Debug("AssemblyLoadFinished: ", assembly_id, " ", hr_status);
+  }
+
+  // keep this lock until we are done using the module,
+  // to prevent it from unloading while in use
+  std::lock_guard<std::mutex> guard(module_id_to_info_map_lock_);
+
+  const auto assembly_info = GetAssemblyInfo(this->info_, assembly_id);
+  if (!assembly_info.IsValid()) {
+    return S_OK;
+  }
+
+  if (debug_logging_enabled) {
+    Debug("AssemblyLoadFinished: AssemblyName=", assembly_info.name);
+  }
+
+  if (assembly_info.name != "Datadog.Trace.ClrProfiler.Managed"_W) {
+    return S_OK;
+  }
+
+  ComPtr<IUnknown> metadata_interfaces;
+  auto hr = this->info_->GetModuleMetaData(assembly_info.manifest_module_id, ofRead | ofWrite,
+                                           IID_IMetaDataImport2,
+                                           metadata_interfaces.GetAddressOf());
+
+  if (FAILED(hr)) {
+    Warn("AssemblyLoadFinished failed to get metadata interface for module id ", assembly_info.manifest_module_id,
+         " from assembly ", assembly_info.name);
+    return S_OK;
+  }
+
+  // Get the IMetaDataAssemblyImport interface to get metadata from the managed assembly
+  const auto assembly_import = metadata_interfaces.As<IMetaDataAssemblyImport>(
+      IID_IMetaDataAssemblyImport);
+  const auto assembly_metadata = GetAssemblyImportMetadata(assembly_import);
+
+  // Configure a version string to compare with the profiler version
+  WSTRINGSTREAM ws;
+  ws << ToWSTRING(assembly_metadata.version.major)
+      << '.'_W
+      << ToWSTRING(assembly_metadata.version.minor)
+      << '.'_W
+      << ToWSTRING(assembly_metadata.version.build);
+
+  // Check that Major.Minor.Build match the profiler version
+  if (ws.str() == ToWSTRING(PROFILER_VERSION)) {
```

Comment:
This is the new assembly version check which, if successful, adds the AppDomainID to our internal `managed_profiler_loaded_app_domains` structure.

Reply:
Is there a better way to do this comparison besides building up a string?

Reply:
What about comparing the `AssemblyMetadata` instance to a string? `PROFILER_VERSION` should always be a `major.minor.build` format, right?

Reply:
The `AssemblyMetadata` class doesn't already have a ToString operation. I noticed that the `Version` struct we use does the same thing that I'm doing here (except that it does a four part assembly version) so this comparison should be fine as-is.